File: static/solutions/cpp/java/src/main/java/com/epi/TransformStringToOther.java
package com.epi;

/*
@slug
transform-one-string-to-other

@summary
Let $s$ and $t$ be strings and $D$ a dictionary, i.e., a set of strings.
Define $s$ to {\em produce} $t$ if there exists a sequence of strings from
the dictionary $P = \langle s_0,s_1,\dots,s_{n-1}\rangle$
such that the first string is $s$, the last string is  $t$, and
adjacent strings have the same length and differ in exactly one character.
The sequence $P$ is called a production sequence.

For example, if the dictionary is $\{\mbox{bat}, \mbox{cot}, \mbox{dog},
\mbox{dag}, \mbox{dot}, \mbox{cat}\}$,
then $\langle \mbox{cat}, \mbox{cot}, \mbox{dot}, \mbox{dog}\rangle$ is
production sequence.

@problem
Given a dictionary $D$ and two strings $s$ and $t$, write a program to determine
if
$s$ produces $t$. Assume that all characters are lowercase alphabets.
If $s$ does produce $t$, output the length of a shortest production sequence;
otherwise, output $-1$.

*/

import java.util.HashSet;
import java.util.LinkedList;
import java.util.Queue;
import java.util.Random;
import java.util.Set;

public class TransformStringToOther {
  private static String randString(int len) {
    Random r = new Random();
    StringBuilder ret = new StringBuilder(len);
    while (len-- > 0) {
      ret.append((char)(r.nextInt(26) + 'a'));
    }
    return ret.toString();
  }

  // @include
  private static class StringWithDistance {
    public String candidateString;
    public Integer distance;

    public StringWithDistance(String candidateString, Integer distance) {
      this.candidateString = candidateString;
      this.distance = distance;
    }
  }

  // Uses BFS to find the least steps of transformation.
  public static int transformString(Set<String> D, String s, String t) {
    Queue<StringWithDistance> q = new LinkedList<>();
    D.remove(s); // Marks s as visited by erasing it in D.
    q.add(new StringWithDistance(s, 0));

    StringWithDistance f;
    while ((f = q.poll()) != null) {
      // Returns if we find a match.
      if (f.candidateString.equals(t)) {
        return f.distance; // Number of steps reaches t.
      }

      // Tries all possible transformations of f.first.
      String str = f.candidateString;
      for (int i = 0; i < str.length(); ++i) {
        String strStart = i == 0 ? "" : str.substring(0, i);
        String strEnd = i + 1 < str.length() ? str.substring(i + 1) : "";
        for (int j = 0; j < 26; ++j) { // Iterates through 'a' ~ 'z'.
          String modStr = strStart + (char)('a' + j) + strEnd;
          if (D.remove(modStr)) {
            q.add(new StringWithDistance(modStr, f.distance + 1));
          }
        }
      }
    }

    return -1; // Cannot find a possible transformations.
  }
  // @exclude

  public static void main(String[] args) {
    Random r = new Random();
    int len;
    if (args.length == 1) {
      len = Integer.parseInt(args[0]);
    } else {
      len = r.nextInt(10) + 1;
    }
    String s = randString(len);
    String t = randString(len);
    Set<String> D = new HashSet<>();
    D.add(s);
    D.add(t);
    int n = r.nextInt(1000000) + 1;
    for (int i = 0; i < n; ++i) {
      D.add(randString(len));
    }
    System.out.println(D);
    System.out.println(s + " " + t + " " + D.size());
    System.out.println(transformString(D, s, t));
  }
}

File: static/python/transform_string_to_other.py
import sys
import random
import string
import collections


def rand_string(length):
    return ''.join(
        random.choice(string.ascii_lowercase) for _ in range(length))


# @include
# Uses BFS to find the least steps of transformation.
def transform_string(D, s, t):
    StringWithDistance = collections.namedtuple(
        'StringWithDistance', ('candidate_string', 'distance'))
    q = collections.deque([StringWithDistance(s, 0)])
    D.remove(s)  # Marks s as visited by erasing it in D.

    while q:
        f = q.popleft()
        # Returns if we find a match.
        if f.candidate_string == t:
            return f.distance  # Number of steps reaches t.

        # Tries all possible transformations of f.candidate_string.
        for i in range(len(f.candidate_string)):
            for c in string.ascii_lowercase:  # Iterates through 'a' ~ 'z'.
                cand = f.candidate_string[:i] + c + f.candidate_string[i + 1:]
                if cand in D:
                    D.remove(cand)
                    q.append(StringWithDistance(cand, f.distance + 1))
    return -1  # Cannot find a possible transformations.
# @exclude


def transform_string_pythonic(D, s, t):
    if s == t:
        return 0
    length = 1
    running = set([s])
    while running:
        running = D & set(cand[:i] + c + cand[i + 1:]
                          for cand in running for i in range(len(cand))
                          for c in string.ascii_lowercase)
        if t in running:
            return length
        length += 1
        D -= running
    return 0


def main():
    length = int(sys.argv[1]) if len(sys.argv) == 2 else random.randint(1, 10)
    s = rand_string(length)
    t = rand_string(length)
    n = random.randint(1, 1000000)
    D = {s, t} | {rand_string(length) for _ in range(n)}
    another_D = D.copy()
    print(s, t, len(D))
    res1 = transform_string(D, s, t)
    res2 = transform_string_pythonic(another_D, s, t)
    print(res1, res2)
    assert res1 == res2


if __name__ == '__main__':
    main()
